### PR TITLE
SI-8030 force symbols on presentation compiler initialization

### PR DESCRIPTION
**Pull request by densh**
This pull request forces a number of built-in symbols in presentation compiler to prevent them from being entered during parsing.

The property “parsing doesn’t enter new symbols” is tested on a rich source file that contains significant number of variations of Scala syntax.

review @retronym

### CODE REVIEW

File: test/files/presentation/parse-invariants/src/a/A.scala [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+  final val x6 = 1
+  lazy final val x7 = 1
+  val Some(x8) = Some(0)
+  var x9 = 1
+  var x10, y10 = 1
+  var x11: Int = 1
+
+  implicit def implicitView: Option[Int] = None
+  def implicitArg1(implicit i: Int) = i + 2
+  def implicitArg2[T: Fooable] = implicitly[Fooable[T]]
```
Comment (retronym) [inline test/files/presentation/parse-invariants/src/a/A.scala:108]: for more variations: view bounds, upper/lower bounds

File: src/interactive/scala/tools/nsc/interactive/Global.scala [hunk 2]
```diff
@@ -1234,6 +1234,16 @@ class Global(settings: Settings, _reporter: Reporter, projectName: String = "")
     }
   }
 
+  protected def forceSymbols(): Unit = {
```
Comment (retronym) [inline src/interactive/scala/tools/nsc/interactive/Global.scala]: `// We force here to ensure that...`

**Comment by retronym** [comment]
@densh  You forgot to commit `Test.scala` it its new location.

**Comment by densh** [comment]
@retronym Yeap, `Test.scala` got lost somewhere. Will amend it tomorrow.

**Comment by densh** [comment]
Just added `Test.scala`, a few examples of bounds on type arguments, and descriptive comment for symbol forcing method.

**Comment by retronym** [comment]
LGTM

File: src/interactive/scala/tools/nsc/interactive/Global.scala [hunk 3]
```diff
@@ -1234,6 +1234,18 @@ class Global(settings: Settings, _reporter: Reporter, projectName: String = "")
...
   }
 
+  // We need to force a number of symbols that might be touched by a parser.
+  // Otherwise thread safety property of parseTree method would be violated.
+  protected def forceSymbolsUsedByParser(): Unit = {
+    val symbols =
+      Set(UnitClass, BooleanClass, ByteClass,
+          ShortClass, IntClass, LongClass, FloatClass,
+          DoubleClass, NilModule, ListClass) ++ TupleClass.seq
+    symbols.foreach(_.initialize)
```
Comment (dotta) [inline src/interactive/scala/tools/nsc/interactive/Global.scala:1244]: @densh Sorry for being late to the party, but why isn't `fullyInitializeSymbol` used in place of the simpler `initialize`? @retronym suggested using `fullyInitializeSymbol` in the related ticket (I'm inlining below his comment):

> And I would change that to `fullyInitializeSymbol` to be on the safe side as you have the tuple class, module, and module class symbols to worry about.

Comment (retronym) [inline src/interactive/scala/tools/nsc/interactive/Global.scala:1244]: It doesn't really matter for the parser, which only needs to refer to the symbols, and doesn't trigger any computations that force the related symbols. The test case here shows this is sufficient.

Comment (dotta) [inline src/interactive/scala/tools/nsc/interactive/Global.scala:1244]: So, we should update https://github.com/scala-ide/scala-ide/pull/595/files. Thanks for clarifying.

Comment (retronym) [inline src/interactive/scala/tools/nsc/interactive/Global.scala:1244]: I don't think it does any harm to initialize the related symbols, though.

Comment (dotta) [inline src/interactive/scala/tools/nsc/interactive/Global.scala:1244]: Right. But keeping the two aligned will allow us to remove the ad-hoc initialization logic from our side once we drop 2.10 support.

**Pull request merged**